### DISCUSSION TITLE
Überdenken der Entität Benennung (z.B. state und lowbat)

### DISCUSSION THREAD

**Discussion by ThomDietrich** [SukramJ/aiohomematic#443] [General]
Hallo,
ich bin gerade frisch auf die neue Integration gewechselt.
Dabei war ich etwas überrascht zu sehen, dass nach wie vor eng mit technischen an Homematic Channels angelehnten Entity Namen gearbeitet wird. Diese sind zweifelsohne technisch eindeutig aber nicht sehr Endanwender-freundlich.

Konkret beziehe ich mich da insbesondere auf `state` und `lowbat`:

![image](https://user-images.githubusercontent.com/2870104/168489051-c9a20410-9320-4fdc-b6f2-035ac5921f9d.png)
![image](https://user-images.githubusercontent.com/2870104/168489126-f63732cf-efbd-4315-b9ac-7e6ca6d79d2b.png)

Die strings "state" bzw "lowbat" kommen sowohl in der entity_id als auch im friendly_name vor. Ich möchte vorschlagen das für beide properties unabhängig zu überdenken! Nicht nur aber insbesondere auch weil das nicht gleichauf mit anderen Integrationen ist.

**Friendly name:** Diese sollten sprechende Namen haben die einem Nutzer intuitiv ein Verständnis der Entity Bedeutung geben - insbesondere durch die UI (z.B. durch die Interpretation als "Normal" bzw. "Low" bei `device_class: battery`).
Weiterhin sollten die Namen aussprechbar sein, bspw zur Interaktion mit Alexa (Aus- und Eingabe).

**Entity_id:** Diese können gern etwas detailierter und technischer sein, aber auch hier könnte aufgeräumt werden.

----

**Vorschlag:** Ich denke, dass der Zusatz "State" völlig unnötig ist. Jede Entität representiert automatisch einen *Zustand* und dieser sollte sich bei vernünftiger Benennung des Geräts automatisch ergeben.
Statt "lowbat" schlage ich "battery" vor. Die Information `device x battery is low` ist deutlich sinniger als `device x lowbat is low`.

Auf meinem System sehen die Entitäten nun so aus:

![image](https://user-images.githubusercontent.com/2870104/168489776-ba3b52cf-8f60-4eb5-b147-599fb8d7644b.png)
![image](https://user-images.githubusercontent.com/2870104/168489788-50dcb11f-4186-4188-b6c5-5fbc1461b1b5.png)

**Comment by danielperna84** [comment]
Ich glaube du stellst dir das zu einfach vor. Im Gegensatz zu alten Integration wird - aus gutem Grund - bei der neuen Integration das meiste automatisch generiert. Die Folge ist, dass Support für neue Geräte nicht erst eingebaut werden muss. Hierbei wird zur Benennung daher der Parameter genommen wie er in der Gerätedefinition steht die von der CCU kommt. Daran wird und sollte sich, meiner Meinung nach, nichts ändern. Ganz zu schweigen davon, dass User immer ihre eigene Definition dessen haben, was ein guter Name für ein Entity ist. Wenn das so gemacht würde wie es dir gefällt, dann kommt nächste Woche jemand der sich darüber beschwert.

**Reply by ThomDietrich** [reply]
Hey Daniel,
dass in der Art gearbeitet wird habe ich gelesen und das finde ich genau richtig! 👏 
Aber sollte nicht genau das alles weitere post-processing erleichtern? Du kannst generisch bleiben und trotzdem eine Logik in dieser Art implementieren:

```python
if (channel_parameter == "LOWBAT" || channel_parameter == "LOW_BAT") {
   parameter = "battery" # o.ä.
}
```

> Daran wird und sollte sich, meiner Meinung nach, nichts ändern.

Den Ansatz finde ich etwas schwierig. Die Gerätedefinition wurde von eq3 für interne technische Zwecke konzipiert und nicht für den Endanwender. Hier sehe ich die Pflicht UND Chance einer Integration wie eurer, sinnvolle sprechende Namen zu wählen, welche auch Anforderungen bezüglich bspw. natürlicher Sprach-Generierung genügen.

> Ganz zu schweigen davon, dass User immer ihre eigene Definition dessen haben, was ein guter Name für ein Entity ist.

Da stimme ich dir zu. Es ist mir in Konsequenz auch nicht wichtig, ob wir am Ende dieser Diskussion bei "battery" oder "low_battery" enden. Die Wahl sollte sinnvoll sein und den Anforderungen von Home Assistant genügen. "state" und "lowbat" tun dies aktuell mit Sicherheit nicht.

Noch ein letzter Gedanke: Natürlich ist diese Integration noch jung und ich sehe an den Tickets, dass euer Fokus noch auf technischeren Themen liegt. Kosmetische Details wie das hier stehen da sicher am Ende der Prioritätenliste, diese sind allerdings wichtig um den Endanwender zu erreichen. Eines der Ziele, welches sich Home Assistant auf die Fahne geschrieben hat.

**Reply by SukramJ** [reply]
Grundsätzlich finde ich den Vorschlag inhaltlich nicht schlecht.
Ich denke etwas weniger Pathos in der Diskussion wäre hilfreich, um das Thema voran zu bringen.

Letztendlich kann jeder Benutzer auch jetzt schon Namen und Entity-Id nach eigenen Vorstellungen über die HA-UI anpassen.

**Reply by ThomDietrich** [reply]
Fair. Usability ist immer etwas Gefühlsangelegenheit aber um der Sache den "Pathos" zu entnehmen:

Ich betreue die Home Assistant Installation eines Blinden Powerusers. Sowohl Sprach-Ausgabe über das iPhone, wie auch Eingabe via Alexa sind mit Namen wie "Wohnzimmer Ecklampe State" nicht direkt möglich.

**Comment by SukramJ** [comment]
Gibt es hierzu eigentlich noch Stimmen aus der Community?
Was halten lang-monatige Benutzer von dem Vorschlag?
@eikowagenknecht @Lemocin @emufan @noxhirsch @Baxxy13
@itn3rd77 @chrste90

**Reply by itn3rd77** [reply]
Finde die Argumente von @ThomDietrich nachvollziehbar und sinnvoll. Ich kann in meiner kleinen Welt natürlich auch ohne überleben. Also wenn Du das machen kannst und willst @SukramJ bringt das die Integration sicherlich weiter voran.

**Reply by eikowagenknecht** [reply]
Ich kann den Vorschlag nachvollziehen und finde ihn gut. Sinnvolle Defaultnamen erleichtern die Bedienung und das Verständnis und sparen Arbeit. "Battery is low" ist viel intuitiver als "Lowbat is on". "State" habe ich als "Status" tatsächlich in einigen Entitäten dringelassen, da es meinem Empfinden nach bei Sensoren ganz gut passt und mir ein oder zwei Namenskonflikte spart. Z.B. gibt es bei mir eg_wohnen_tuer_status -> Türsensor und eg_wohnen_tuer -> Türschloss zum Öffnen. Ist aber sicher Geschmackssache.

Die Defaultwerte sind mir persönlich (also für Einsatz bei mir zu Hause) allerdings nicht sonderlich wichtig. Ich benenne ohnehin sämtliche Entitäten nach Ersteinrichtung um nach dem Schema <Geschoss>.<Raum> <Gerät> <Funktion> und dabei auch die Entitäts ID analog (nur mit Unterstrichen statt "." und " ") und da jede Integration ihre eigene Konvention verfolgt (so jedenfalls meine Erfahrung) und die HA Entitätsgenerierung Umlaute seltsam ersetzt bin ich eh dauernd am Umbenennen. Per Script geht das ganz gut, habe ich hier geposted: https://community.home-assistant.io/t/bulk-edit-for-entities/289593/5

Für Alexa lege ich auch für jedes Gerät per YAML einen benutzerdefinierten Namen für die Steuerung fest, da Alexa ein anderes Namensschema hat als meine Home Assistant Entitäten, um die Standardräume etc. zu nutzen dort ("Wohnzimmer Stehlampe" statt "EG.Wohnen Hue Stehlampe"). Ist ein wenig extra Arbeit, aber sooo oft richtet man zum Glück ja keine neuen Geräte ein.

Das nur als kurzer Einblick, wie andere damit umgehen.

**Reply by danielperna84** [reply]
@SukramJ 

Wie sieht es denn mit der "Lokalisierbarkeit" aus? Kann der String der dann am Ende genommen wird aus den Translations in `custom_homematic` kommen? Wir haben ja jetzt schon mindestens DE und NL als Zielsprache, von EN als Default mal abgesehen. Ich meine HM-Geräte werden auch in Italien und Polen verkauft, und in der Schweiz spricht man ja auch nicht nur Deutsch. Daher müssten / könnten umgeschriebene Parameter sich ja an der Sprache orientieren in der HA betrieben wird.

**Reply by noxhirsch** [reply]
Da kann ich mich nur anschließen - kurz gesagt: Ich finde den Vorschlag nachvollziehbar und sehr sinnvoll, und finde, dass es die Integration verbessern würde. Aber es wäre auch nicht super dramatisch, wenn es nicht umgesetzt würde, wenn es (auch langfristig betrachtet) mit unverhältnismäßigem Aufwand verbunden wäre.

**Reply by SukramJ** [reply]
@danielperna84 
Im Zusammenhang mit diesem Thema hab ich auch mal wieder nach Lokalisierbarkeit für Entitätsnamen gesucht aber bei HA nichts gefunden. Auch nicht im Architektur-Repo.
Wenn man das selber machen will braucht man eine Liste aller Parameter (Extract aus Pydevccu) und entsprechende Übersetzungen für unterstützte Sprachen ([Lokalize](https://lokalise.com/) unterstützt teilweise kostenfrei OpenSource Projekte). Der entsprechende Sprache müsste natürlich vom Frontend geliefert werden.

In dieser Diskussion geht es aber erst mal um:
- den Ersatz des technischen Names (LOW_BAT) auf einen natürlicher Namen (Battery).
- Entfernen von nicht notwendigen Namensbestandteilen (STATE, LEVEL)
Die Implementierung hierfür ist nicht aufwendig.

In einem weiteren Schritt würde dann eine Lokalisierung greifen, die sich dann immer auf Entitätsnamen und ID auswirkungen hätte.
Die Implementierung hierfür ist aufwendiger, und es gibt auch viele Aufgaben im Umfeld.

**Reply by emufan** [reply]
Ich wäre eher dabei, dass es so genommen wird, wie es ist und bei Daniels ersten Kommentar oben. Das Mapping ist dann ja ansonsten komplett undurchschaubar für den Anwender, der in CCU, Device-Doku und dem Ergebnis dieser Integration dann die Dinge nicht mehr zusammenbringt. Und dann kommt HA und normalisiert die Status nach Devicetypen, usw. auch nochmal. Wem die Friendly-Names und Namen nicht gefallen, der hat sie ja auch schon in der Vergangenheit geändert.

Darüber hinaus (je nachdem ob nur ehemalige Attribute "normalisiert" werden oder auch die Hauptkanäle, erschwert das auch die Migration von alt auf neu, da bislang viele Haupt Entity IDs gleich sind wie vorher, ggf. sonst nicht mehr.

**Reply by SukramJ** [reply]
Wie wichtig ist die Migration?

Die Änderung betrifft:
STATE:
- switch: Bidcos-RF Switch (Erklärung [unten](https://github.com/danielperna84/hahomematic/discussions/443#discussioncomment-2776619))
- binary_sensor: Kontakt-, Fenster-Sensoren

alle LOW_BAT / LOWBAT

**Reply by emufan** [reply]
low_bat: Etliche Automatisierungen und zwei Dashboards. Ich persönlich kann per Studio Code Server drübergehen, Suche&Ersetzen, was aber auch nur geht, wenn der sonstige Aufbau bleibt und sich nur das Suffix ändert., Aber nicht jeder traut sich an die storage-Dateien.

binary_sensor: Dito, nur halt viel mehr Skripte, Automoatisierungen und Dashboards. Und zusätzlich die Thematik von Eiko, dass ich hier schon andere Entitäten gleichen Namens aus anderen Integrationen habe, ohne "state" dran. Dann bekomme ich schon mal initial (2)er. Usw.

Ich kann damit leben und umbenennen. Aber das Thema, wie man möglichst leicht von der alten zur neuen kommt, hatten wir ja auch schon öfter und das zieht einen neuen manuellen Schritt ein, der vorher nicht war.

BTW ein low_bat Boolean ist für mich auch weniger als eine Batterie. Darunter verstehe ich (und versteht HA mE sonst) ein numerischen 0-100% Ding. Suggeriert durch eine Umbenennung mE mehr als es in Wahrhheit war und ist.

**Reply by SukramJ** [reply]
Eigentlich denke ich , das es sich bei der Änderung **nicht** um einen Breaking Change handelt.
Jede Entität hat im Hintergrund eine `unique_id`, die ich bei dieser Umstellung nicht anpassen würde.
Die `entity_id` wird bei der **erstmaligen** Anlage in HA aus dem Namen abgeleitet und in der `entity_registry` gespeichert.
Der Name wird aber neu übernommen, so das die Änderung auch bei bestehenden Installationen sichbar wird.

Nur bei neuen Installationen werden auch die `entity_ids` nach neuem Schema erstellt.

**Reply by emufan** [reply]
Das von mir praktizierte Vorgehen und das was ich bislang jedem empfehlen würde, ist aber:

- Alte Integration löschen.
- Alle alten Entititäten löschen
- Resultat: Automatisierungen, Skripte und UIs sind alle noch da mit alter entity_id
- Neustart
- Neue Integration einrichten
- '> 80% der alten entity_ids sind wieder so wie vorher
- Resultat  Automatisierungen, Skripte und UIs funktionieren ohne Anpassunge wieder und weiter

die unique_id sieht der Anwender doch gar nicht und Du beschreibst ja nur die technischen Details. Die stimmen zwar, aber obiges Vorgehen ist halt einfach praktisch.

Ah, vielleicht reden wir auch aneinander vorbei. Ich spreche nicht von Migration von low_bat zu battery bei eingerichteter neuen Integration, sondern von der alten HM-Integration zur neuen hier.

**Reply by SukramJ** [reply]
Und das machst Du mit jedem release update der custom_component?
Ich Rede vom release Update und nicht von einer Migration.

**Reply by emufan** [reply]
Siehe letzter Absatz: Mein Punkt ist die kommenden einmalige Migration bei vielen vielen Nutzern von der alten zur neuen Integration. Die bei state, low_bat, usw. nichts tun müssten, mit einer Normalisierung aber schon. Und dort die (Mehrzahl?), die nicht in den Textdateien rumwurschteln und das ggf. im UI suchen und ändern, ...

**Reply by SukramJ** [reply]
Ok, dann musst Du deine Aufzählung nur ein wenig anpassen:

'> 70% der alten entity_ids sind wieder so wie vorher

Das die entity_ids Alt --> Neu relativ gut passen ist purer Zufall.
Ich hab versucht die neue Integration sinnvoll zu gestalten und nicht den Migrationspfad einfach zu gestalten.

Bei einer einmaligen Tätigkeit wie einer Migration halte ich den Mehraufwand für vertretbar.
Ich habe bei den Änderungen in erster Linie die aktuellen Nutzer der custom_component im Blick.

**Reply by emufan** [reply]
Ich will hier nicht um Zahlen feilschen. Ob Zufall oder nicht. Hatte vorher geklappt und passt hier dann nicht mehr. Ich hab's schon hinter mir und ich könnte mir helfen, wenn nicht. Wollte nur anmerken, dass es das Thema gibt.

**Reply by emufan** [reply]
... und das battery statt low_bat wirklich semantisch falsch wäre. Eine Batterie ist nicht an/aus, bzw. das an/aus sagt nicht das aus, das low_bat bedeutet. Hier sollte es zumindest einen treffenderen Namen geben.

**Reply by SukramJ** [reply]
Ein Dilemma:
Zusammen mit der device_class `battery` (On means low, Off means normal) macht es in der UI schon Sinn.
In Automatisierung hantiert man allerdings mit on und off, dann sieht es wieder nicht so sinnvoll aus.
Bei `low_bat` ist es eher anders herum.

Welche alternativ Vorschläge gibt es für `Battery`?

**Reply by ThomDietrich** [reply]
@emufan es gibt leider keine mir bekannte offizielle Vorgabe für battery Entitäten aber eine gute Richtung gibt das dafür populäre Blueprint vor: https://gist.github.com/sbyx/1f6f434f0903b872b84c4302637d0890 -> ein `binary_sensor` mit `device_class`"battery" ist okay.

> Ich hab versucht die neue Integration sinnvoll zu gestalten und nicht den Migrationspfad einfach zu gestalten.

@SukramJ Freut mich das zu lesen ;)

Übersetzung: Ich weiß leider auch nicht welche Optionen bzgl Lokalisierung von Entities zur Verfügung stehen. Ich erinnere mich nicht, dass bei einer anderen Integration schon gesehen zu haben, geht also vermutlich aktuell nicht. Ich würde vorschlagen zunächst bei englischen Begriffen zu bleiben aber ich quatsch die Tage mal mit einem befreundeten Core Entwickler darüber. Vielleicht gibt es Pläne der Art.

> Ein Dilemma

Dieses Dilemma gibt es bei allen device_classes. On/Off ist auch bei z.B. Fensterkontakten nicht sehr intuitiv. Imho ein Thema welches Home Assistant systematisch lösen muss. In der UI ist dies bereits geschafft und auch in Automatisierungen sollte das triggern auf "low" statt "on" möglich werden. Ebenso eine Frage die ich mal ansprechen werde.

**Reply by emufan** [reply]
> ein `binary_sensor` mit `device_class`"battery" ist okay.

Hab' ich auch kein Problem mit. Mag nur keine Pferde anmalen und sagen, es wären Zebras. 😉 Gegen das default-Verhalten von device_class battery (On means low, Off means normal) sträubt sich meine Intuition ebenso. Wobei die Implementierung auch bei meinen Geräten noch nicht gesehen habe, sodass es mich (noch) nicht betrifft. Aber vielleicht nutzen andere Integrationen die device_class bei binary-Sensoren auch mit einem sinnvollen Namen, wie low_bat oder so. 😎

Für mich ist eine battery die Batterie des Geräts an sich und ist voll/funktionsfähig oder leer.
Und das HM Indikator ein Indikator, ob die Batterie des Geräts leer ist (true) oder nicht leer (false).

Unabhängig, dass ich es aus den anderen genannten Gründen persönlich nicht anders nennen würde, als in HM, so würde ich persönlich halt nur für einen erwartbaren, semantisch richtiger Namen nebst State-Verhalten plädieren. Und da gefällt mir persönlich halt battery so gar nicht, da es halt das Gegenteil ist, es ist keine Batterie, sondern ein Batterie_leer_Indikator.

**Comment by Baxxy13** [comment]
Also den ersten Punkt "Ersatz des technischen Namens auf einen natürlicheren Namen" finde ich gut. Das macht ja die WebUI mit ihren ganzen "Translations" auch. 

Beim Punkt "Entfernen von nicht notwendigen Namensbestandteilen" bin ich etwas zwiegespalten. 
Erscheint mir irgendwie inkonsistent wenn einige Entitäten den Datenpunkt nicht als Bestandteil haben.
Aktuell habe ich ein:
`HmIP-STHD Dev2_514:1 - Bedienung Actual Temperature`
und ein:
`HM-ES-PMSw1-Pl LEQ0529290 State`

(Ich war zu faul das Testgerät in der WebUI umzubenennen)
Lasse ich jetzt das "State" weg...
`HM-ES-PMSw1-Pl LEQ0529290`
...hat die Entität den gleichen Namen wie das Gerät:
`HM-ES-PMSw1-Pl LEQ0529290`
(k.A. ob das problematisch ist)

Also wenn automatisch was weggelassen werden soll, dann müsste m.E. geprüft werden ob der zum Datenunkt gehörige Kanal nicht den default-Namen hat.

**Reply by SukramJ** [reply]
Guter Einwand, aber aktuell sind die Namen schon inkonsitent.
Bsp.:
Bei einer Custom_Entity wie einem `HmIP-PSM` wird bei der Hauptentität (Parameter STATE) der Parameter gar nicht zur Namensfindung hinzugenommen. Alle HmIP-Switche sind Custom_Entities, um die virtuellen Kanäle auszublenden.
Bei einer Generic_Entity wie einem `HM-ES-PMSw1-Pl` wird der Parameter immer zur Namensfindung hinzugenommen. Alle Bidcos-RF-Switche sind Generic_Entities.
Also statt switch.buecherregal_state heisst die Entitaet dann switch.buecherregal.

Gerätename und eine Entitaet können auch ohne Probleme gleich heissen.

**Reply by ThomDietrich** [reply]
Hey @Baxxy13,
schade dass du deinem Gerät nicht vorher via WebUI einen Namen gegeben hast. Das ist hier nämlich entscheidend. Wenn der Gerätename sinnvoll gesetzt ist brauchst du "state" nicht. Siehe mein Beispiel aus dem ersten Beitrag.

Alternatives Beispiel von @eikowagenknecht:

> "State" habe ich als "Status" tatsächlich in einigen Entitäten dringelassen, da es meinem Empfinden nach bei Sensoren ganz gut passt und mir ein oder zwei Namenskonflikte spart. Z.B. gibt es bei mir eg_wohnen_tuer_status -> Türsensor und eg_wohnen_tuer -> Türschloss zum Öffnen. Ist aber sicher Geschmackssache.

Die Beispiele ergeben Sinn und es gibt sicher immer die Ausnahme in der eine manuelle Anpassung wertvoll ist. Bei mir hätten diese Geräte aber andere Namen und "state" ist nicht länger erforderlich: `eg_wohnen_tuer` und `eg_wohnen_tuerschloss`. Die UI zeigt dann sprechend "EG Wohnen Tuer - Offen" bzw. "EG Wohnen Türschloss - Entriegelt" und eine Regel klappt auch: `WHEN eg_wohnen_tuer is closed THEN verschließe(eg_wohnen_tuerschloss)`
